### PR TITLE
Use DB in the backend

### PR DESCRIPTION
**Pull request by carlosms**
Implements #41.

Please take a look @dpordomingo, @smacker.

TODOs:

- [X] Update README with new env variable (done in e23160c)
- [x] Check if DB strings need to be escaped

### CODE REVIEW

File: cli/server/server.go
Comment (smacker) [inline cli/server/server.go]: does it look weird to pass DB connection to a router only for me?
(no changes required, just asking)

File: server/dbutil/dbutil.go
Comment (smacker) [inline server/dbutil/dbutil.go]: what do you think about composition here?
It would allow not write `db.SQLDB.method` everywhere, just `db.method`

File: cli/server/server.go
Comment (carlosms) [reply cli/server/server.go]: I'm ok with moving the repositories initialization back to the server.go file. I just did that to avoid having lots of arguments here. But of course we could always store all the repos in a single struct...

Let me know if I should change this.

File: server/handler/assignments.go
Comment (smacker) [inline server/handler/assignments.go]: I see we do it often. Just an idea for the future:

```go
func URLParamInt(r, key) (int, error) {
  str := chi.URLParam(r, "pairId")
  return strconv.Atoi(str)
}
```

File: server/handler/auth.go
Comment (smacker) [inline server/handler/auth.go]: I noticed this pattern with passing fields of structs a few times already.
No changes are required now, but I prefer to pass a model to a repository.
It makes life easier when the model starts growing.

File: server/repository/assignments.go
Comment (smacker) [inline server/repository/assignments.go]: according to https://github.com/src-d/code-annotation/issues/2 it should be `nil`. Frontend also expects `nil` there.

File: server/handler/auth.go
Comment (carlosms) [reply server/handler/auth.go]: Makes sense. But the problem, in this specific case, is that there is a field of the User model that is not up to you, the ID will be chosen by the DB.

You could create a User model and leave the ID as 0, but that can lead to bugs in my opinion... you may be tempted to continue using the `User` object you created. In the current code it is clear for the caller that the complete `User` object will be provided by `Create`.

Any other opinions?

File: server/repository/file_pairs.go
Comment (smacker) [inline server/repository/file_pairs.go]: maybe it should be private?

File: cli/server/server.go
Comment (smacker) [reply cli/server/server.go]: no, please don't. It simplifies the code.

File: server/handler/assignments.go
Comment (smacker) [inline server/handler/assignments.go]: we do it often. How about moving it to a helper?
```
func checkUserInCtx(ctx) (id, error)
```

File: server/repository/assignments.go
Comment (carlosms) [reply server/repository/assignments.go]: We could use a null value in the DB, but isn't it going to end up as an empty string when we read it back?
https://github.com/src-d/code-annotation/pull/52/files#diff-b01fac8859e3c130c7e1a433072473edR24

Comment (smacker) [reply server/repository/assignments.go]: no if you use [sql.NullString](https://golang.org/pkg/database/sql/#NullString)

Comment (carlosms) [reply server/repository/assignments.go]: Ok... what about in serializers.go, how does the final json field look like?

Comment (smacker) [reply server/repository/assignments.go]: it should be `*string`

File: server/handler/auth.go
Comment (smacker) [reply server/handler/auth.go]: Sorry I don't get the problem with ID.

```
u := User{login,name,...}
err := repo.save(&u)
fmt.Println(u.ID) // prints user id
```

keep id as zero value for objects that aren't saved yet is totally normal.

You can see in for example in [most popular ORM for go](https://github.com/jinzhu/gorm)

**Comment by dpordomingo** [comment]
@carlosms please, ping me once you address the Maxim feedback to perform a second review :D

**Comment by dpordomingo** [comment]
@carlosms your work looks great,
please, ping me once you address the Maxim feedback to perform a second review :D

File: server/repository/file_pairs.go
Comment (carlosms) [reply server/repository/file_pairs.go]: DB is initialized in router.go. Or should it be a private field with a setter method?

File: server/dbutil/dbutil.go
Comment (carlosms) [reply server/dbutil/dbutil.go]: Done in 77f23a1

File: server/handler/assignments.go
Comment (carlosms) [reply server/handler/assignments.go]: done in 01f1cbc

File: server/handler/auth.go
Comment (carlosms) [reply server/handler/auth.go]: done in b09114c

File: server/handler/assignments.go
Comment (carlosms) [reply server/handler/assignments.go]: done in c9d26ae

File: server/repository/assignments.go
Comment (carlosms) [reply server/repository/assignments.go]: done in 07e2f9c

File: server/repository/assignments.go
Comment (smacker) [inline server/repository/assignments.go]: defer rows.Close()

File: server/repository/assignments.go
Comment (smacker) [inline server/repository/assignments.go]: defer rows.Close()

File: server/repository/file_pairs.go
Comment (smacker) [reply server/repository/file_pairs.go]: usually, people do it with a constructor

**Review by smacker — CHANGES_REQUESTED** [review]

File: server/repository/assignments.go
Comment (carlosms) [reply server/repository/assignments.go]: done in e3886d2

File: server/handler/render.go
Comment (dpordomingo) [inline server/handler/render.go]: s/URLParamInt/urlParamInt
Anyhow, private functions use to be undocummented.

File: server/handler/render.go
Comment (dpordomingo) [inline server/handler/render.go]: We're trying to avoid named returns.
The code below this comment can be replaced by:
```go
if err != nil {
    err = serializer.NewHTTPError(
        http.StatusBadRequest,
        fmt.Sprintf("Wrong format for URL parameter %q; received %q", key, str)
    )
}

return val, err;
```

It is only needed to change the returned because `val` is already `0` when an error occurs.

File: server/router.go
Comment (dpordomingo) [inline server/router.go]: It is not saved a `FilePair` but an `Assignment`, that's because it should be:
`s/pairId/assignmentId`

File: server/handler/assignments.go
Comment (dpordomingo) [inline server/handler/assignments.go]: It is not saved a `FilePair` but an `Assignment`, that's because it should be:
`s/pairId/assignmentId`

File: server/handler/assignments.go
Comment (dpordomingo) [inline server/handler/assignments.go]: It is not saved a `FilePair` but an `Assignment`, that's because it should be:
`s/pairId/assignmentId`
>_We considered that the same `FilePair` can be assigned to the same `User` many times in different experiments, so it can not be `PRIMARY KEY`_

File: server/model/models.go
Comment (dpordomingo) [inline server/model/models.go:34]: What `Score` is? We didn't consider that in our initial schema...
Do you have any doc about it?

File: server/dbutil/dbutil.go
Comment (dpordomingo) [inline server/dbutil/dbutil.go]: We considered that the same `FilePair` can be assigned to the same `User` many times in different experiments, so it can not be `PRIMARY KEY`

File: server/handler/user.go
Comment (dpordomingo) [inline server/handler/user.go]: `404` only if `!error` and `!user`
```go
if err != nil {
    return nil, err
}
if user == nil {
    return nil, serializer.NewHTTPError(http.StatusNotFound, "user not found")
}
```

File: server/repository/assignments.go
Comment (dpordomingo) [inline server/repository/assignments.go]: Can the insertion succeed, but the `Get` fail? And if so, does it mean that the creation failed? I don't think so...
At this point, I'd replace the if and everything below this comment with:
```go
return err
```

File: server/repository/assignments.go
Comment (dpordomingo) [inline server/repository/assignments.go]: This can return many assignments:
>_We considered that the same `FilePair` can be assigned to the same `User` many times in different experiments, so it can not be `PRIMARY KEY`_

File: server/repository/assignments.go
Comment (dpordomingo) [inline server/repository/assignments.go]: `s/pairId/assignmentId`

File: server/repository/assignments.go
Comment (dpordomingo) [inline server/repository/assignments.go]: >_We considered that the same `FilePair` can be assigned to the same `User` many times in different experiments, so it can not be `PRIMARY KEY`_

The idea of checking the `user_id` is redundant if it is used the `assignmentId`, but it can be a good idea to ensure that the assignment being saved, belongs to the user in session.

File: server/repository/experiments.go
Comment (dpordomingo) [inline server/repository/experiments.go]: _idem_
Can the insertion succeed, but the `Get` fail?
```go
return err
```

File: server/repository/users.go
Comment (dpordomingo) [inline server/repository/users.go]: _idem_
Can the insertion succeed, but the `Get` fail?
```go
return err
```

File: server/repository/experiments.go
Comment (dpordomingo) [inline server/repository/experiments.go]: Do we assume this method is going to be needed in the future?
I can not imagine with which purpose... Any idea?

If it makes sense to need it, I'd rename by `GetByName` (same as `GetById`)

File: server/service/oauth.go
Comment (dpordomingo) [inline server/service/oauth.go]: I think it does not need to be public, because handlers does not build it, just use it

**Review by dpordomingo — CHANGES_REQUESTED** [review]
Wow! that's great, I only found some things (repeated in some places), here is the list:
- one `FilePair` can be assigned to the same `User` many times in different experiments, so it can not be `PRIMARY KEY`
- avoid named returns
- use http status `500` when user can not be fetched
- what `FilePair.Score` is?
- the error returned by `repository.Whatever.Create(Whatever)` must be the the error returned by the `insert` query
- do we really need `repository.Experiment.GetByName(string)?

File: server/dbutil/dbutil.go
Comment (smacker) [reply server/dbutil/dbutil.go]: don't think currently it's a big problem. The problem is FE expects to have an id for assignment because it's standard in REST.

File: server/repository/experiments.go
Comment (smacker) [reply server/repository/experiments.go]: I would propose to avoid `Not implemented` methods at all as long as we don't need them.

File: server/dbutil/dbutil.go
Comment (carlosms) [reply server/dbutil/dbutil.go]: According to [the schema](https://github.com/src-d/code-annotation/issues/2#issuecomment-357315069), a FilePair belongs to one and only one experiment.

File: server/handler/assignments.go
Comment (carlosms) [reply server/handler/assignments.go]: This makes sense assuming the current primary key of the assignments table is correct.
Right now there isn't an assignment ID (again, following  [the schema](https://github.com/src-d/code-annotation/issues/2#issuecomment-357315069)).

If this looks bad, we could actually change the route to be `PUT /experiments/{experimentId}/`, and get hte pairID from the body.

File: server/model/models.go
Comment (carlosms) [reply server/model/models.go:34]: Score is the similarity score, it comes from the imported data.
It was added when real data was provided by the ML team, see [here](https://github.com/src-d/code-annotation/issues/2#issuecomment-359045024).

File: server/handler/assignments.go
Comment (smacker) [reply server/handler/assignments.go]: please add `assignmentId ` to schema to follow rest conventions and don't rewrite frontend (I believe it's easier to add id, than change FE)

File: server/dbutil/dbutil.go
Comment (dpordomingo) [reply server/dbutil/dbutil.go]: mmm... I did't fond where it was written that a FilePair belongs to only one experiment. It's true that the oposite was not written too, but I feel it's more compatible with the user necessities.

File: server/repository/experiments.go
Comment (dpordomingo) [reply server/repository/experiments.go]: I think that non implemented methods are a good way to define the architecture and the API, but only when it is really possible to need them. But in this case, I don't see when are we using `repo.Experiment.GetByName`

Comment (smacker) [reply server/repository/experiments.go]: it's exactly what I mean. If we don't see where we are going to use them soon ("don't need" in my previous comment), let's avoid them.

File: server/dbutil/dbutil.go
Comment (carlosms) [reply server/dbutil/dbutil.go]: Well, in the schema the file pairs table has an experiment ID column, I understood that as a relationship with one and only one experiment.

But as we discussed via chat, both of you you agree that is a requirement, so I'll change it to be a * to * relationship in separate table, and perform a join when selecting all file pairs belonging to an experiment in Assignments.Initialize

**Comment by bzz** [comment]
@carlosms to speed up the reviews, could you please summarize the current status of the PR? 

Asking, as In PR description there are some TODO items that are not checked and then if I'm not mistaken, there seems to be some un-addressed feedback. 

If addressing the reviews requires substantial time, I find very useful putting together a [comment with markdown checkboxes](https://github.com/src-d/code-annotation/pull/16#issuecomment-358992502), to visualize the progress better and this way to communicate the status with reviewers.

**Comment by carlosms** [comment]
@bzz the description checkboxes were TODOs I found myself to remind me of doing that when the final version before the merge was ready.

- [X] Update README with new env variable (done in e23160c)
- [x] Check if DB strings need to be escaped

The current un-adressed feedback are the following two changes:

- [x] Add an assignment ID as the primary key of the assignments table
- [ ] ~~Allow file-pairs to belong to several experiments~~ New issue: https://github.com/src-d/code-annotation/issues/66

File: server/repository/users.go
Comment (carlosms) [reply server/repository/users.go]: Even if it's unlikely, the Get may fail if the DB connection breaks, or someone deletes the entry at that very exact moment.
In that case the user pointer is not correctly populated with the assigned ID, and I'd say the method should notify of this returning an error.

**Comment by carlosms** [comment]
Do you @smacker & @dpordomingo think we can leave the requirement 'Allow file-pairs to belong to several experiments' out of this PR?

I propose to open an issue to do it later it in a different PR once this branch is in master. Reasoning:
- It affects code that was already in master
- it's not really needed anytime soon, we actually only have one hardcoded experiment.
- this way we can merge this PR and #56 as soon as possible.

File: server/handler/assignments.go
Comment (smacker) [inline server/handler/assignments.go]: 1. `StatusForbitten` may be better here. It's a common approach to `logout` user on `StatusUnauthorized` when there is a separate frontend application. Because you usually return this error in auth middleware. (for example when token is outdated)
2. is it better to `Get(id, userID)`?

not a blocker though

File: server/repository/file_pairs.go
Comment (smacker) [inline server/repository/file_pairs.go]: something weird is going on here. You never use `a`&`b` beside scanning values to them.

File: server/repository/users.go
Comment (smacker) [inline server/repository/users.go]: do you need this get to get `id`?
this way it's not going to work. Simplified example of what you are doing: https://play.golang.org/p/dlEmRgTL1uN

File: server/service/jwt.go
Comment (smacker) [inline server/service/jwt.go]: `return ctx.Value(userIDKey).(int)`

File: server/repository/file_pairs.go
Comment (carlosms) [reply server/repository/file_pairs.go]: Good catch, this should be pair.Left and Right. I missed it because the serializer does nothing with these two values.

File: server/repository/users.go
Comment (dpordomingo) [reply server/repository/users.go]: mmm... I see two patterns:
- `error := userRepo.Create(user)` -> it should not modify the user, but return an error if it failed storing in the DB
- `error := user.Save()` -> it stores the user, and assigns its ID

If we go with the first one, the ID should be assigned from the outside of the Create method, I mean: instead of hiding the `.Get` call inside the `.Create` method, doing it from the outside:
```go
if error := userRepo.Create(userToStore); err != nil { 
    return err
}
// in some cases, we won't need to continue, so the Get call won't be
// neccessary, but if we need a user with ID, we can continue:
user, err := userRepo.Get(user.Login)
if err != nil { 
    return err
}
//at this point, we have a valid user
fmt.Println(user)
```
^ would you approve that @smacker ?

I saw that many times to avoid hiding complexity in methods that seem to be "low level" (and `userRepo.Create` seems to be that kind of methods)

Comment (dpordomingo) [reply server/repository/users.go]: *PS:*
if it is assumed a possible race condition and the necessity of calling `.Get` from the `.Create` to modify the `user` retrieved as an input parameter, it can happen unexpected things;
It can happen that the final (modified) `user` will be quite different than the input one (another Name, Login...) so the caller function will operate with that `user` without noticing it was "changed in behind" by the `.Create` func.

(that's another reason because it could be a bad idea to modify the input user inside the `.Create` func)

**Review by smacker — APPROVED** [review]

**Review by dpordomingo — APPROVED** [review]
LGTM; you did a great work @carlosms  
Many thanks!!!

I have only one concern about the final implementation as explained in the comment https://github.com/src-d/code-annotation/pull/52#discussion_r166358719 and https://github.com/src-d/code-annotation/pull/52#discussion_r166381084, but if @smacker disagrees I can stand without it.

The thinks you marked as "TODO", imo can be done in a separated PR, and they're not blocker:
- [ ] Update README with new env variable -> just agree with @smacker how to push to #29
- [ ] Allow file-pairs to belong to several experiments -> your current implementation seems to follow our specs, and it works; it could be done in the future if really needed.

Before merging I'd only require:
- [ ] open an issue explaining the (possible) problem with unescaped things,
- [ ] rebase and squash to obtain a history that makes sense to you.

File: server/repository/users.go
Comment (smacker) [reply server/repository/users.go]: I'm okay with current solution and your proposal is okay too. Both approaches are common. For error in `get` and not in `exec` people invented transactions if we really need it.

I would suggest to merge it as is now.

**Comment by dpordomingo** [comment]
Since @smacker is ok with both approaches https://github.com/src-d/code-annotation/pull/52#discussion_r166388220 lets merge when @bzz approves the PR and: 
- [ ] open an issue explaining the (possible) problem with unescaped things, if necessary,
- [ ] rebase and squash to obtain a history that makes sense to you.

**Comment by bzz** [comment]
Both, handling `'Allow file-pairs to belong to several experiments' out of this PR` and adding usage of the transactions for complex failure modes in subsequent PRs makes sense to me.

Great job @carlosms !

Probably a nitpick but 2 things:

 - [x] One thing I find useful is to have string literal extracted to the constants inside the code.
   The reasons are:
   - it makes SQL queries more visible
   - they are grouped in the single place
   - it simplifies the subsequent refactoring 
   - and facilitates re-use (easier to notice if you already have it defined)

 - [x] Another minor but nice practice is
    In case you are using `//TODO` to label some short-term subsequent fixes is to save some time for a readers by adding `//TODO(gh-username)` so no need for `git glame` while looking at it later.

Would you mind adding those?

One very last and very minor thing, but for me, as a reviewer, it's a bit confusing to merge PR with todo checkboxes not checked in PR description/comments of PR author - they are communicating the idea of more work needed to be done. It would help to avoid the confusion if you could update https://github.com/src-d/code-annotation/pull/52#issue-293184299 and  https://github.com/src-d/code-annotation/pull/52#issuecomment-363378896 . 

Feel free to use ~strikethough~ for items that are intended not to be checked in this PR, as they have a link to elsewhere, where this item is been taken care of.

**Review by bzz — APPROVED** [review]
LGTM.

Thank you Carlos!

**Comment by carlosms** [comment]
Replying @dpordomingo 
> open an issue explaining the (possible) problem with unescaped things, if necessary,

With the final code strings are inserted using the `$N` query placeholders, which should be safe. The only place where a "raw" string is used is [here](https://github.com/src-d/code-annotation/pull/52/files#diff-ebc27875bd59733c7382e0a9013ff999R131), but the string is validated to be one of `model.Answers`. So I won't open an issue, no work is needed.

In reply to @bzz 
> Both, handling 'Allow file-pairs to belong to several experiments' out of this PR

New issue created: https://github.com/src-d/code-annotation/issues/66

> and adding usage of the transactions for complex failure modes in subsequent PRs makes sense to me.

Are you refering to this comment https://github.com/src-d/code-annotation/pull/52#discussion_r166388220 ? I don't think a transaction is the proposed solution, as @smacker suggested to leave the current code as it is.

**Comment by bzz** [comment]
Sounds good to me. Let's :shipit:

**Pull request merged**
